[DTensor] Make DTensor `from_local` backward partial() to replicate() pass through

yoyoyocmu opened the pull request:
Summary:
This change makes the `DTensor.from_local()` placements in backward pass from `Partial()` to `Replicate()` as pass through for following reasons:
1. When we run backward pass of DTensor.from_local, if the target placement is partial() (i.e. from user manual overwrite code instead of torch_dispatch) we keep the grad as replicate. This is because converting the gradients back to `Partial()` is meaningless.
2. The current div logic will lead to wrong numerical value in the above case.

Test Plan:
**CI**:
CI Tests

**Unit test**:
`buck2 test mode/dev-nosan //caffe2/test/distributed/_tensor:redistribute`
- Passed

**With model training**:
```
# We tested the case where input tensor is manually overwrite as Partial() and
# output tensor manually overwrite to Shard() then to local.

# Before the change: numerical value not correct
Forward pass:
    collective: ReduceScatter
backward pass:
    collective: AllGather + div by process group size


# After the change: div is removed as expected.
Forward pass:
    collective: ReduceScatter
Backward pas:
    collective: AllGather
```

Differential Revision: D52175709




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

wanchaol commented on test/distributed/_tensor/test_redistribute.py:
```diff
@@ -175,6 +181,30 @@ def test_replicate_to_partial(self):
+            (Replicate(),),
+        )
+
+        out_local_forward = redistribute_local_tensor(
```
Let's not test this API as it's sort of a private API, instead we just directly test `dtensor.from_local + redistribute`? I think it's already been tested in the above test too. If you want to just add some more assertions about the sizes of the gradients like you did here, we can remove the test here.

yoyoyocmu replied:
Good catch! Removed this test in the latest version.

wanchaol reviewed (commented):
change looks good, have some additional comments inlined

wanchaol commented on torch/distributed/_tensor/redistribute.py:
```diff
@@ -76,6 +76,7 @@ def redistribute_local_tensor(
     local_tensor: torch.Tensor,
     current_spec: DTensorSpec,
     target_spec: DTensorSpec,
+    is_backward: bool = False,
```
I guess this can be done in a follow up PR as it involves changing Shard -> Partial too

yoyoyocmu replied:
Will have another PR to do the refactor.

yoyoyocmu committed:
[DTensor] Make DTensor `from_local` backward partial() to replicate() pass through (#115967)

Summary:

This change makes the `DTensor.from_local()` placements in backward pass from `Partial()` to `Replicate()` as pass through for following reasons:
1. When we run backward pass of DTensor.from_local, if the target placement is partial() (i.e. from user manual overwrite code instead of torch_dispatch) we keep the grad as replicate. This is because converting the gradients back to `Partial()` is meaningless.
2. The current div logic will lead to wrong numerical value in the above case.

Test Plan:
**CI**:
CI Tests

**Unit test**:
`buck2 test mode/dev-nosan //caffe2/test/distributed/_tensor:redistribute`
- Passed

**With model training**:
```
# We tested the case where input tensor is manually overwrite as Partial() and
# output tensor manually overwrite to Shard() then to local.

# Before the change: numerical value not correct
Forward pass:
    collective: ReduceScatter
backward pass:
    collective: AllGather + div by process group size


# After the change: div is removed as expected.
Forward pass:
    collective: ReduceScatter
Backward pas:
    collective: AllGather
```

Reviewed By: wanchaol

Differential Revision: D52175709
